### ISSUE TITLE
PyTorch Inference Mode

### ISSUE THREAD

**Issue by mj-will** [mj-will/nessai#79]
An inference mode was added to PyTorch in [version 1.9.0](https://pytorch.org/blog/pytorch-1.9-released/#beta-inference-mode-api). It seems to be an even more extreme version of `torch.no_grad()` and promises to be even faster.

To quote the documentation [here](https://pytorch.org/docs/master/notes/autograd.html#locally-disable-grad-doc):

>Inference mode is the extreme version of no-grad mode. Just like in no-grad mode, computations in inference mode are not recorded in the backward graph, but enabling inference mode will allow PyTorch to speed up your model even more.

We should consider testing inference mode in the *population* stage of `nessai` since this requires repeated calls to the normalising flow and is the main bottleneck at the moment.

**Useful links**

[PyTorch 1.9.0 release notes](https://pytorch.org/blog/pytorch-1.9-released/#beta-inference-mode-api)
[Inference mode documentation](https://pytorch.org/docs/stable/generated/torch.inference_mode.html)
[Explanation of different autograd mechanics](https://pytorch.org/docs/master/notes/autograd.html)
